python/pyspark/streaming/dstream.py

```diff
@@ -51,125 +78,170 @@ class DStream:
         """
 
-        def func(iterator):
-            return filter(f, iterator)
+        def func(iterator: Iterable[T_co]) -> Iterable[U]:
+            return filter(f, iterator)  # type: ignore[arg-type]
 
         return self.mapPartitions(func, True)
 
-    def flatMap(self, f, preservesPartitioning=False):
+    def flatMap(
+        self: "DStream[T_co]",
+        f: Callable[[T_co], Iterable[U]],
+        preservesPartitioning: bool = False,
+    ) -> "DStream[U]":
         """
         Return a new DStream by applying a function to all elements of
         this DStream, and then flattening the results
         """
 
-        def func(s, iterator):
+        def func(s: int, iterator: Iterable[T_co]) -> Iterable[U]:
             return chain.from_iterable(map(f, iterator))
 
         return self.mapPartitionsWithIndex(func, preservesPartitioning)
 
-    def map(self, f, preservesPartitioning=False):
+    def map(
+        self: "DStream[T_co]", f: Callable[[T_co], U], preservesPartitioning: bool = False
+    ) -> "DStream[U]":
         """
         Return a new DStream by applying a function to each element of DStream.
         """
 
-        def func(iterator):
+        def func(iterator: Iterable[T_co]) -> Iterable[U]:
             return map(f, iterator)
 
         return self.mapPartitions(func, preservesPartitioning)
 
-    def mapPartitions(self, f, preservesPartitioning=False):
+    def mapPartitions(
+        self, f: Callable[[Iterable[T_co]], Iterable[U]], preservesPartitioning: bool = False
+    ) -> "DStream[U]":
         """
         Return a new DStream in which each RDD is generated by applying
         mapPartitions() to each RDDs of this DStream.
         """
 
-        def func(s, iterator):
+        def func(s: int, iterator: Iterable[T_co]) -> Iterable[U]:
             return f(iterator)
 
         return self.mapPartitionsWithIndex(func, preservesPartitioning)
 
-    def mapPartitionsWithIndex(self, f, preservesPartitioning=False):
+    def mapPartitionsWithIndex(
+        self,
+        f: Callable[[int, Iterable[T_co]], Iterable[U]],
+        preservesPartitioning: bool = False,
+    ) -> "DStream[U]":
         """
         Return a new DStream in which each RDD is generated by applying
         mapPartitionsWithIndex() to each RDDs of this DStream.
         """
         return self.transform(lambda rdd: rdd.mapPartitionsWithIndex(f, preservesPartitioning))
 
-    def reduce(self, func):
+    def reduce(self, func: Callable[[T_co, T_co], T_co]) -> "DStream[T_co]":
         """
         Return a new DStream in which each RDD has a single element
         generated by reducing each RDD of this DStream.
         """
-        return self.map(lambda x: (None, x)).reduceByKey(func, 1).map(lambda x: x[1])
+        return (
+            self.map(lambda x: (None, x))  # type: ignore[misc]
+            .reduceByKey(func, 1)  # type: ignore[arg-type]
+            .map(lambda x: cast(T_co, x[1]))
```

Comment:
As above, you shouldn't need ignores here if you switch to non-covariant typevars.